### PR TITLE
Refactor non-streaming generate

### PR DESCRIPTION
**Pull request by xwu99**
- Convert str to list before generate
- Define types for generate input and streamline preprocess_prompts
- Move all predictor classes to predictors directory

### CODE REVIEW

File: llm_on_ray/inference/predictors/hpu_predictor.py [hunk 1]
```diff
@@ -179,8 +185,17 @@ def get_streamer(self):
                 self.tokenizer, skip_prompt=True, timeout=0, skip_special_tokens=True
             )
 
-    def generate(self, prompt, **config):
+    # FIXME: support MultiplePromptInput and MllmPromptInput
+    def generate(self, input: GenerateInput, **config) -> GenerateOutput:
+        if not isinstance(input, SinglePromptInput):
+            raise TypeError(
+                "HPUPredictor doesn't support multiple prompts and MLLM prompts for now!"
+            )
```
Comment (kira-lin) [inline llm_on_ray/inference/predictors/hpu_predictor.py:193]:
```suggestion
        if isinstance(input, MllmPromptInput):
            raise TypeError(
                "HPUPredictor doesn't support MLLM prompts for now!"
            )
```

File: llm_on_ray/inference/predictors/hpu_predictor.py [hunk 2]
```diff
@@ -179,8 +185,17 @@ def get_streamer(self):
                 self.tokenizer, skip_prompt=True, timeout=0, skip_special_tokens=True
             )
 
-    def generate(self, prompt, **config):
+    # FIXME: support MultiplePromptInput and MllmPromptInput
```
Comment (kira-lin) [inline llm_on_ray/inference/predictors/hpu_predictor.py:188]:
```suggestion
    # FIXME: support MllmPromptInput
```

File: llm_on_ray/inference/predictors/hpu_predictor.py [hunk 3]
```diff
@@ -57,9 +58,14 @@
...
 from llm_on_ray.inference.inference_config import (
     InferenceConfig,
-    GenerateResult,
+    ModelGenerateResult,
+)
+from llm_on_ray.inference.predictor import (
+    GenerateInput,
+    GenerateOutput,
+    Predictor,
+    SinglePromptInput,
```
Comment (kira-lin) [inline llm_on_ray/inference/predictors/hpu_predictor.py:67]:
```suggestion
    SinglePromptInput,
    MultiplePromptInput,
    MllmPromptInput,
```

**Review by kira-lin — CHANGES_REQUESTED** [review]

File: llm_on_ray/inference/predictors/transformer_predictor.py [hunk 4]
```diff
@@ -102,33 +112,28 @@ def streaming_generate(self, prompt, streamer, **config):
...
-                    input_length=input_length,
-                    generate_length=gen_tokens.size()[1] - input_length,
-                )
-                for i in range(len(prompts))
-            ]
-
-        return None
+            for i in range(len(prompts))
+        ]
+        return results[0] if isinstance(prompts, SinglePromptInput) else results
```
Comment (KepingYan) [inline llm_on_ray/inference/predictors/transformer_predictor.py:136]: Should it be ‘input’ here? Because prompts always is a List here.

Comment (xwu99) [reply llm_on_ray/inference/predictors/transformer_predictor.py:136]: thanks, good catch!

File: llm_on_ray/inference/predictors/mllm_predictor.py [hunk 5]
```diff
@@ -89,32 +94,37 @@ def _process_config(self, config):
...
             **inputs,
             stopping_criteria=self.stopping_criteria,
             streamer=streamer,
             **config,
         )
 
-    def generate(self, image, prompt, **config):
+    def generate(self, input: GenerateInput, **config) -> GenerateOutput:
+        if not isinstance(input, tuple):
+            raise TypeError("MllmPredictor should use (image, prompt) as input.")
```
Comment (kira-lin) [inline llm_on_ray/inference/predictors/mllm_predictor.py:115]:
```suggestion
            raise TypeError("MllmPredictor should use (prompt, image) as input.")
```

**Review by kira-lin — COMMENTED** [review]
LGTM

**Review by kira-lin — APPROVED** [review]

**Pull request merged**